newIDE/app/src/ResourcesList/ResourceSelector.js

```diff
@@ -163,32 +163,71 @@ export default class ResourceSelector extends React.Component<Props, State> {
+      resourcesLoader,
+      resourceKind,
+    } = this.props;
     const { resourceName } = this.state;
-
-    const resourceNames = [];
-    if (project.getResourcesManager().hasResource(resourceName)) {
-      resourceNames.push(resourceName);
+    const resourcesManager = project.getResourcesManager();
+    const initialResource = resourcesManager.getResource(resourceName)
+    
+    let initialResourceMetadata = {};
+    const initialResourceMetadataRaw = initialResource.getMetadata();
+    if (initialResourceMetadataRaw) {
+      try {
+        initialResourceMetadata = JSON.parse(initialResourceMetadataRaw);
+      } catch(e) { console.error("Malformed metadata", e); }
+      // console.log('resource metadata found, sending it to editor:');
+      // console.log(initialResourceMetadata)
     }
 
-    resourceExternalEditor.edit({
-      project,
-      resourcesLoader,
-      singleFrame: true,
-      resourceNames,
-      extraOptions: {
-        fps: 0,
-        name: 'Image',
-        isLooping: false,
-      },
-      onChangesSaved: resources => {
-        if (!resources.length) return;
+    let externalEditorOptions = {};
+    if (resourceKind === 'image') {
+      const resourceNames = [];
+      if (resourcesManager.hasResource(resourceName)) {
+        resourceNames.push(resourceName);
+      }
+      externalEditorOptions = {
+        project,
+        resourcesLoader,
+        singleFrame: true,
+        resourceNames,
+        extraOptions: {
+          fps: 0,
+          name: 'Image',
+          isLooping: false,
+        },
+        onChangesSaved: resources => {
+          if (!resources.length) return;
 
-        // Burst the ResourcesLoader cache to force images to be reloaded (and not cached by the browser).
-        resourcesLoader.burstUrlsCacheForResources(project, [resources[0].name]);
-        this.props.onChange(resources[0].name);
-      },
-    });
+          // Burst the ResourcesLoader cache to force images to be reloaded (and not cached by the browser).
+          resourcesLoader.burstUrlsCacheForResources(project, [
+            resources[0].name,
+          ]);
+          this.props.onChange(resources[0].name);
+        },
+      };
+    } else if (resourceKind === 'audio') {
+      let initialResourcePath = resourcesLoader.getFullUrl(project, resourceName)
+      initialResourcePath = initialResourcePath.substring(
+        7,
+        initialResourcePath.lastIndexOf('?cache=')
+      );
+      externalEditorOptions = {
+        project,
+        extraOptions: {
+          initialResourcePath,
+          initialResourceMetadata,
+        },
+        onChangesSaved: (newResourceData,newResourceName) => {          
+          this.props.onChange(newResourceName);
+          const newResource = resourcesManager.getResource(newResourceName);
+          newResource.setMetadata(JSON.stringify(newResourceData[0].metadata));
+        },
+      };
+    };
+    resourceExternalEditor.edit(externalEditorOptions);
```

Comment:
So if you look at the flow error, you can see this:
![image](https://user-images.githubusercontent.com/1280130/47119288-3bff1180-d262-11e8-83b3-4fe7e23507d1.png)

What Flow is telling us is that there is a chance that here we are calling `resourceExternalEditor.edit ` with an object that is missing a loooot of properties. How is that possible? Well, if both conditions `resourceKind === 'audio'` and `resourceKind === 'image'` are false. 
Flow is wise to tell us this because this could happen in the future :)
The solution is simple => Either add a `else { return }` so that Flow understand that we handle the case. Even better, simply duplicate the call `resourceExternalEditor.edit(externalEditorOptions);` inside the two if branches and bonus: use const instead of let for `externalEditorOptions` ;)

Reply:
I was wondering about that. Thank you for the patience and for explaining how flow does these checks. I didn't realize that it was quite that sophisticated in the way it analyzes the code

Reply:
Yeah it's quite sophisticated. Can hurt a bit at the beginning with cryptic error messages, but it's still evolving and once you get use to it, it's a invaluable tool for the long term maintainability/stability of the app.